### DISCUSSION TITLE
How to resolve `You are trying to rewrite {N} public commits` error?

### DISCUSSION THREAD

**Discussion by andresoto-x** [arxanas/git-branchless#988] [Q&A]
Im running into the following issue when I run `git restack`. I have a chain of branches and was modifying one of the intermediate branches. 

```
git restack
You are trying to rewrite 7 public commits, such as: a0e262d GraphQL CIM API to support types from the Domain package. (#1832)
It is generally not advised to rewrite public commits, because your
collaborators will have difficulty merging your changes.
Retry with -f/--force-rewrite to proceed anyways.
```

**Comment by andresoto-x** [comment]
Steps to reproduce. 

```bash
$ git sl 
◇ 72381e6 1h (main)
$ echo test > test
$ git add . 
$ git commit --amend --no-edit

$ git sl
◇ abc1234 6m (main) main description

$ git checkout -b foo
$ echo foo > foo
$ git add . 
$ git commit --m "foo"

$ git sl
◇ abc1234 6m (main) main description
┃
◯ 3ceab8a 4m (foo) foo

$ git checkout -b foo2
$ echo foo2 > foo2
$ git add .
$ git commit -m "foo2"
$ git sl

◇ abc1234 6m (main) main description
┃
◯ 3ceab8a 4m (foo) test
┃
● a49bbd9 25s (ᐅ foo2) foo2

$ git co main
$ git pull
$ git sl
⋮
◇ 987654 6m commit before main
┣━┓
⋮ ✕ abc1234 6m (rewritten as 123456) main description
⋮ ┃
⋮ ◯ 3ceab8a 4m (foo) test
⋮ ┃
⋮ ◯ a49bbd9 25s (ᐅ foo2) foo2
⋮
◆ 123456 4s (ᐅ main) main description

$ git restack
You are trying to rewrite 1 public commit, such as: 123456 main description
It is generally not advised to rewrite public commits, because your
collaborators will have difficulty merging your changes.
```

**Reply by arxanas** [reply]
Something seems to be wrong with your `git pull` unrelated to git-branchless. As demonstrated in your last smartlog, the old version of `main` pointed to `abc1234`, but the new version of `main` points to `123456`. Normally, after a `git pull`, the old `main` commit will be an ancestor of the new `main` commit. Since in your case it isn't, if you were to immediately `git push`, then it should be rejected because the new `main` is not a descendant of the old `main`. If you were to force-push, then you would be rewriting public history, which is what git-branchless is warning about.

If this is okay in your case, you can of course run `git restack` with the `-f`/`--force-rewrite` flag to ignore the warning and rebase your branches on top of the new `main`, but the principled fix is to resolve whatever your `git pull` is doing so that the resulting `main` is a descendant of its previous value.

You can also try `git restack 'stack()'` to see if the warning disappears. This became the new default in a recent version of git-branchless (and I'm not sure if it was released in a public release yet).

**Reply by claytonrcarter** [reply]
@andresoto-x I don't understand something about your example. Can you clarify? In your example, there are 4 commits related to main: 
- `987654 commit before main`
- `72381e6 (no description)` ← this is `main` but also detached HEAD when you start your example
- `abc1234 main description` ← this is the new `main` and still a detached  HEAD commit after you run `git commit --amend --no-edit` (3rd command)
- `123456 main description` ← this is `main` (which is checked out) after you run `git pull`

So, in your example, has new work been added to the remote `main`? Or is your local `main` "up to date" when you start your example? If it's up to date, I'm surprised that that `git pull` succeeded since it would seem that `main` would have divergent states locally vs remote (though I'm not very familiar w/ working on divergent branches like that).

Also, can you clarify what you would expect to happen?
